### ISSUE TITLE
Allow Result file path customization or templating

### ISSUE THREAD

**Issue by anna-geller** [PrefectHQ/prefect#7212]
### First check

- [X] I added a descriptive title to this issue.
- [X] I used the GitHub search to find a similar request and didn't find it.
- [X] I searched the Prefect documentation for this feature.

### Prefect Version

2.x

### Describe the current behavior

Currently, users can set `PREFECT_LOCAL_STORAGE_PATH` but the Results are stored as UUID so that they never collide

Each file gets stored as `PREFECT_LOCAL_STORAGE_PATH/uuid` where uuid is a JSON file:

```json
{
    "serializer": {
        "type": "pickle",
        "picklelib": "cloudpickle",
        "picklelib_version": "2.1.0",
    },
    "data": "gAWVMQAAAAAAAACMLUhpIGZyb20gMTNfYWx3YXlzX3N0b3JlX3Jlc3VsdHNfbG9jYWxseSEg8J+R\ni5Qu\n",
    "prefect_version": "2.6.1",
}
```

## Problem
The issue with this approach is that it's **hard to match the Result file to the flow run** and/or task run that generated that result, which is useful for troubleshooting.

### Describe the proposed behavior

We may set the default Result location to `None`, which would store the result as `PREFECT_LOCAL_STORAGE_PATH/uuid` as usual

We could expose templating that could be set on the profile settings or on a flow/task decorator. If this is set to some value (not `None`), then Result would be stored in a (Jinja?)-templated location e.g.:

```
PREFECT_DEFAULT_RESULT_LOCATION='${PREFECT_LOCAL_STORAGE_PATH}/
{date.year}/{date.month}/{date.day}/{flow_name}/{flow_run_id}/{task_run_name}.json'
```

### Example Use

Troubleshooting of failed runs -- see what was the last result before the flow run failed

**Comment by thomasfrederikhoeck** [comment]
I think it would be a sensible default to at least have `flow_name` and `flow_run_id` as part of the path :-)

**Comment by madkinsz** [comment]
Resolved by #8924
